### PR TITLE
Improve deleting a src file UX

### LINKED ISSUE #22: Review the experience of deleting a source file
Deleting a source file requires currently a few interactions.

Run
```
$ cd example
$ nix-shell --pure --run "bazel run //:gazelle"
$ rm package-c/multiple-folder-src/PackageC/Exposed/E.hs
```
The first error we get is
```
$ nix-shell --pure --run "bazel run //:gazelle"
...
gazelle: himportscan: /home/facundo/tweag/gazelle_haskell_modules/example/package-c/multiple-folder-src/PackageC/Exposed/E.hs: openFile: does not exist (No such file or directory)
gazelle: exit status 1
```
which could be improved to point at the rule that has the missing file.

After removing the `haskell_module` rule that points at this file, `gazelle` succeeds.
```
$ nix-shell --pure --run "bazel test //..."
...
ERROR: /home/facundo/tweag/gazelle_haskell_modules/example/package-c/BUILD.bazel:57:16: in modules attribute of _haskell_library rule //package-c:package-c-multiple-srcs: target '//package-c:package-c-multiple-srcs.PackageC.Exposed.E' does not exist. Since this rule was created by the macro 'haskell_library', the error might have been caused by the macro implementation
```
After removing `:package-c-multiple-srcs.PackageC.Exposed.E` from the modules attribute of `package-c-multiple-srcs` we get
```
$ nix-shell --pure --run "bazel test //..."
...
ERROR: /home/facundo/tweag/gazelle_haskell_modules/example/package-c/BUILD.bazel:168:16: HaskellBuildObject //package-c:check-package-c-multiple-srcs //package-c:check-package-c-multiple-srcs.CheckPackageCMultipleSrcs failed: (Exit 1): ghc_wrapper failed: error executing command bazel-out/host/bin/external/rules_haskell/haskell/ghc_wrapper ... (remaining 2 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox

package-c/checksrc/CheckPackageCMultipleSrcs.hs:4:1: error:
    Could not find module ‘PackageC.Exposed.E’
...
  |
4 | import PackageC.Exposed.E ()
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
which finally requires fixing the code.

All of this steps multiplied by several removed files are rather tiresome.
The `fix` command should detect the removed files and update the configuration.

### PR DESCRIPTION
**Pull request by googleson78**
Fixes #22

### CODE REVIEW

File: gazelle_haskell_modules/lang.go [hunk 1]
```diff
@@ -215,10 +214,47 @@ func (*gazelleHaskellModulesLang) Fix(c *config.Config, f *rule.File) {
...
 
+func cleanupModulesList(r *rule.Rule, ruleNameSet map[string]bool) {
+	// TODO: use labels instead of manually stripping away the ':' ?
+	stripColon := func(module string) string { return module[1:] }
+	cleanupModulesLists(r, ruleNameSet, "modules", stripColon)
+}
+
+func cleanupHiddenModulesList(r *rule.Rule, ruleNameSet map[string]bool) {
+	ruleName := r.Name()
+	// TODO: use something better?
```
Comment (googleson78) [inline gazelle_haskell_modules/lang.go:232]: Not sure if this isn't already implemented somewhere around here, but I didn't find it.

File: gazelle_haskell_modules/lang.go [hunk 2]
```diff
@@ -215,10 +214,47 @@ func (*gazelleHaskellModulesLang) Fix(c *config.Config, f *rule.File) {
...
 		}
+		cleanupModulesList(r, ruleNameSet)
+		cleanupHiddenModulesList(r, ruleNameSet)
 	}
+
 	f.Sync()
 }
 
+func cleanupModulesList(r *rule.Rule, ruleNameSet map[string]bool) {
+	// TODO: use labels instead of manually stripping away the ':' ?
```
Comment (googleson78) [inline gazelle_haskell_modules/lang.go:225]: Is there a better way to do this?

**Comment by googleson78** [comment]
TODO: add fix invocation to CI

File: himportscan/src/HImportScan/ImportScanner.hs [hunk 3]
```diff
@@ -60,8 +61,15 @@ instance Aeson.ToJSON ModuleImport where
 -- | Retrieves the names of modules imported in the given
 -- source file. Runs the GHC lexer only as far as necessary to retrieve
 -- all of the import declarations.
-scanImportsFromFile :: FilePath -> IO ScannedImports
-scanImportsFromFile filePath = scanImports filePath <$> Text.readFile filePath
+-- If a file is missing, we return 'Nothing'.
+-- TODO: It would be better to give more information on the missing file,
```
Comment (googleson78) [inline himportscan/src/HImportScan/ImportScanner.hs:65]: I would do this here, but I would need to alter the JSON format with which go and hs communicate, so I'd prefer this to be done in a separate PR.

File: himportscan/exe/Main.hs [hunk 4]
```diff
@@ -5,11 +5,14 @@ module Main where
...
 import qualified Data.Aeson as Aeson
 import qualified Data.ByteString.Lazy.Char8 as ByteString.Lazy
+import Data.Maybe (catMaybes)
 
 main :: IO ()
 main = do
   haskellSrcs <- lines <$> getContents
-  mapM scanImportsFromFile haskellSrcs >>= printImports
+  -- silently ignore missing files
+  -- TODO: it would be better to instead report them to the user
```
Comment (googleson78) [inline himportscan/exe/Main.hs:14]: I would do this here, but I would need to alter the JSON format with which go and hs communicate, so I'd prefer this to be done in a separate PR.

File: gazelle_haskell_modules/rule_generation.go [hunk 5]
```diff
@@ -119,15 +119,19 @@ func haskellModuleRulesToRuleInfos(
...
 
-		ruleInfoss = append(ruleInfoss, []*RuleInfo{&ruleInfo})
+			ruleInfoss = append(ruleInfoss, []*RuleInfo{&ruleInfo})
 
-		r.SetPrivateAttr(PRIVATE_ATTR_MODULE_NAME, ruleInfo.ModuleData.ModuleName)
-		r.SetPrivateAttr(PRIVATE_ATTR_ORIGINATING_RULE, ruleInfo.OriginatingRules)
+			r.SetPrivateAttr(PRIVATE_ATTR_MODULE_NAME, ruleInfo.ModuleData.ModuleName)
+			r.SetPrivateAttr(PRIVATE_ATTR_ORIGINATING_RULE, ruleInfo.OriginatingRules)
+		} else if len(modDatas) == 0 {
+			log.Printf("found no module data for haskell_module %s\n", r.Name())
```
Comment (googleson78) [inline gazelle_haskell_modules/rule_generation.go:133]: Perhaps this should also suggest running fix?

File: gazelle_haskell_modules/lang.go [hunk 6]
```diff
@@ -225,6 +227,13 @@ func fixModulesList(r *rule.Rule, ruleNameSet map[string]bool) {
 	fixModulesLists(r, ruleNameSet, "modules", stripColon)
 }
 
+func fixHiddenModulesList(r *rule.Rule, ruleNameSet map[string]bool) {
+	ruleName := r.Name()
+	// TODO: use something better?
+	addPackageName := func(module string) string { return fmt.Sprintf(":%s.%s", ruleName, module) }
```
Comment (facundominguez) [inline gazelle_haskell_modules/lang.go:233]: IIUC, you want to drop the colon from the output.

File: gazelle_haskell_modules/lang.go [hunk 7]
```diff
@@ -213,26 +213,37 @@ func (*gazelleHaskellModulesLang) Fix(c *config.Config, f *rule.File) {
...
+}
+
+// Leaves only those modules from r that are in ruleNameSet according to toRuleName.
 // If a module is not in ruleNameSet, that means that there was no info associated with it,
 // hence it should be deleted.
-func fixModuleLists(r *rule.Rule, ruleNameSet map[string]bool) {
-	modules := r.AttrStrings("modules")
+//
+// modulesFieldName is expected to be "modules" or "hidden_modules".
+func fixModulesLists(r *rule.Rule, ruleNameSet map[string]bool, modulesFieldName string, toRuleName func(string) string) {
```
Comment (facundominguez) [inline gazelle_haskell_modules/lang.go]: Maybe rename `toRuleName` as `attrStringToRuleName`.

File: example/fix-cleanup-modules/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,69 @@
+load("@rules_haskell//haskell/experimental:defs.bzl", "haskell_module")
```
Comment (facundominguez) [inline example/fix-cleanup-modules/BUILD.bazel:1]: This BUILD file could have a comment explaining what is being tested and how.

**Review by facundominguez — COMMENTED** [review]
Thanks @googleson78!

Very nice breakdown of commits.

I added some comments. The PR is already in pretty good shape though.

File: gazelle_haskell_modules/lang.go [hunk 9]
```diff
@@ -213,10 +213,29 @@ func (*gazelleHaskellModulesLang) Fix(c *config.Config, f *rule.File) {
...
+// Leaves only those modules from r that are in ruleNameSet.
+// If a module is not in ruleNameSet, that means that there was no info associated with it,
+// hence it should be deleted.
+func fixModuleLists(r *rule.Rule, ruleNameSet map[string]bool) {
+	modules := r.AttrStrings("modules")
+	if modules != nil {
+		nonDeletedModules := make([]string, 0, len(modules))
+		for _, module := range modules {
+			// TODO: use labels instead of manually stripping away the ':' ?
+			if ruleNameSet[module[1:]] {
```
[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go]: In the latest version of the code perhaps something like this would do:
```diff
-       stripColon := func(module string) string { return module[1:] }
+       stripColon := func(module string) string {
+               if strings.HasPrefix(module, ":") {
+                       return module[1:]
+               } else {
+                       return module
+               }
+       }
```

[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go]: I tried a fix in 8e5ab77

File: gazelle_haskell_modules/rule_generation.go [hunk 10]
```diff
@@ -119,15 +119,19 @@ func haskellModuleRulesToRuleInfos(
 		}
 
 		modDatas := haskellModulesToModuleData([]string{src})
-		ruleInfo := RuleInfo{
-			OriginatingRules: oRules,
-			ModuleData:       modDatas[0],
-		}
+		if len(modDatas) == 1 {
```
[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/rule_generation.go]: I implemented this in b2f66f2

File: gazelle_haskell_modules/rule_generation.go [hunk 11]
```diff
@@ -119,15 +119,19 @@ func haskellModuleRulesToRuleInfos(
...
 
-		ruleInfoss = append(ruleInfoss, []*RuleInfo{&ruleInfo})
+			ruleInfoss = append(ruleInfoss, []*RuleInfo{&ruleInfo})
 
-		r.SetPrivateAttr(PRIVATE_ATTR_MODULE_NAME, ruleInfo.ModuleData.ModuleName)
-		r.SetPrivateAttr(PRIVATE_ATTR_ORIGINATING_RULE, ruleInfo.OriginatingRules)
+			r.SetPrivateAttr(PRIVATE_ATTR_MODULE_NAME, ruleInfo.ModuleData.ModuleName)
+			r.SetPrivateAttr(PRIVATE_ATTR_ORIGINATING_RULE, ruleInfo.OriginatingRules)
+		} else {
+			log.Printf("warning: found no module source for haskell_module rule %s: %q\n", r.Name(), src)
```
Comment (facundominguez) [inline gazelle_haskell_modules/rule_generation.go:133]: I enhanced this message in a49dd86. I finally didn't add a suggestion to use the fix command to reduce the noise in the output and keep the code simpler. This message as it stands is printed both when fixing and when not fixing.

File: gazelle_haskell_modules/lang.go [hunk 12]
```diff
@@ -213,26 +213,37 @@ func (*gazelleHaskellModulesLang) Fix(c *config.Config, f *rule.File) {
...
+	fixModulesLists(r, ruleNameSet, "modules", stripColon)
+}
+
+// Leaves only those modules from r that are in ruleNameSet according to toRuleName.
 // If a module is not in ruleNameSet, that means that there was no info associated with it,
 // hence it should be deleted.
-func fixModuleLists(r *rule.Rule, ruleNameSet map[string]bool) {
-	modules := r.AttrStrings("modules")
+//
+// modulesFieldName is expected to be "modules" or "hidden_modules".
```
[the remark this answers is not in the document]
Comment (googleson78) [reply gazelle_haskell_modules/lang.go:247]: Upon trying to applies this, I feel like it makes the code less clear. It is true that we're doing string manipulation here, but we still need to keep in mind these are specific kinds of strings. I also agree that calling the passed thing "modules" is a bit misleading, since they need some string manipulation to become the names of module rules.

Is there some middle ground between these two?

File: gazelle_haskell_modules/lang.go (hunk 6 above)
Comment (googleson78) [reply gazelle_haskell_modules/lang.go:239]: What do you mean by "output"? I'm adding this here so it matches the format in `ruleNameSet`. It seems a bit brittle though - is there some way to make it better?

File: gazelle_haskell_modules/lang.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (googleson78) [reply gazelle_haskell_modules/lang.go]: Actually, I think that right now (and before the patch too), if a user writes a fully qualified bazel target in `modules`, it will get deleted, regardless of whether the module exists or not, because `ruleNameSet` won't have the fully qualified module name.

So there probably needs to be something more complex here - either store full target names in `ruleNameSet`, or do a more complex check here.

I think the fully qualified names might be a better solution, as this string manipulation stuff feels very brittle both here and for `hidden_modules` :/

File: gazelle_haskell_modules/rule_generation.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (googleson78) [reply gazelle_haskell_modules/rule_generation.go]: Thanks. I wonder if we should (in another issue perhaps) change the logic here to error if we receive `len(modData) > 1`, as an assert, since it shouldn't be possible?

File: gazelle_haskell_modules/lang.go (hunk 6 above)
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go]: I stripped the colon in 5496a59.

File: gazelle_haskell_modules/lang.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go:247]: The rationale for my suggestion is that the code doesn't care what the string is, as long as the output of toRuleName is a rule name that might appear in the set. But I don't feel strong about it. We can keep the names as they are.

File: gazelle_haskell_modules/lang.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (googleson78) [reply gazelle_haskell_modules/lang.go]: https://github.com/tweag/gazelle_haskell_modules/commit/8e5ab7701e2811c3709608fbd8fcf93b896780e7 

Seems to have broken CI - need to investigate.

[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go]: > I think the fully qualified names might be a better solution, as this string manipulation stuff feels very brittle both here and for hidden_modules :/

A question that helps determining the severity of issues in general is: what the best workaround is?
In light of this question, you are correct that absolute labels in `modules` can confuse the fix command, but it is also easy for the user to make the labels relative instead, and relative labels are the most economical to use anyway. So it is unlikely to be a barrier for adoption.

File: example/fix-cleanup-modules/BUILD.bazel (hunk 8 above)
Comment (facundominguez) [reply example/fix-cleanup-modules/BUILD.bazel:1]: Fantastic!

File: gazelle_haskell_modules/lang.go (hunk 7 above)
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go]: Won't do. See https://github.com/tweag/gazelle_haskell_modules/pull/24#discussion_r861721786

File: gazelle_haskell_modules/rule_generation.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/rule_generation.go]: I'd be fine with such a change.

File: gazelle_haskell_modules/lang.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go]: > Seems to have broken CI - need to investigate.

Mea culpa :see_no_evil: 
Fixed in efae1e7

File: gazelle_haskell_modules/lang.go (hunk 6 above)
Comment (googleson78) [reply gazelle_haskell_modules/lang.go]: I'm not really sure why it works both with and without the colon (this code is out of cache right now :sweat_smile: ) but I trust your judgement :)

File: gazelle_haskell_modules/rule_generation.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (googleson78) [reply gazelle_haskell_modules/rule_generation.go]: I'll open an issue after we merge this PR.

File: gazelle_haskell_modules/lang.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go]: And still I was dropping non-relative labels. I tried another fix in e18e2de.

**Comment by facundominguez** [comment]
@googleson78, are there any other changes that you would like to have related to this PR?

File: gazelle_haskell_modules/lang.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go:245]: I added this bit as part of e18e2de
```
// The field is removed if all the strings are dropped from the list.
```

File: gazelle_haskell_modules/lang.go (hunk 6 above)
Comment (facundominguez) [reply gazelle_haskell_modules/lang.go]: A first short answer is that ruleNameSet doesn't include the leading `:`, so it needs to be removed. Though note that the code has evolved since this conversation. Happy to engage with more details if needed.

File: gazelle_haskell_modules/rule_generation.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (googleson78) [reply gazelle_haskell_modules/rule_generation.go]: #28

[the remark this answers is not in the document]
Comment (googleson78) [reply gazelle_haskell_modules/rule_generation.go]: #29
